File: mockups/serve.mjs
// Zero-dependency static server for the Babun2 mockups folder.
// Serves files from ./mockups on port 4321; "/" → clients-current.html.
import { createServer } from 'node:http';
import { readFile } from 'node:fs/promises';
import { join, normalize, extname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { dirname } from 'node:path';

const ROOT = dirname(fileURLToPath(import.meta.url));
const PORT = 4321;

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.webp': 'image/webp',
  '.ico': 'image/x-icon',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
};

const server = createServer(async (req, res) => {
  try {
    let pathname = decodeURIComponent(new URL(req.url, `http://localhost:${PORT}`).pathname);
    if (pathname === '/' || pathname === '') pathname = '/clients-current.html';

    // Prevent path traversal: resolve within ROOT only.
    const filePath = normalize(join(ROOT, pathname));
    if (!filePath.startsWith(ROOT)) {
      res.writeHead(403, { 'content-type': 'text/plain; charset=utf-8' });
      res.end('Forbidden');
      return;
    }

    const data = await readFile(filePath);
    const type = MIME[extname(filePath).toLowerCase()] || 'application/octet-stream';
    res.writeHead(200, { 'content-type': type, 'cache-control': 'no-cache' });
    res.end(data);
  } catch (err) {
    res.writeHead(err.code === 'ENOENT' ? 404 : 500, { 'content-type': 'text/plain; charset=utf-8' });
    res.end(err.code === 'ENOENT' ? 'Not found' : 'Server error');
  }
});

server.listen(PORT, () => {
  console.log(`Babun2 mockups server running at http://localhost:${PORT}/`);
});

File: mockups/serve.py
#!/usr/bin/env python3
# Zero-dependency static server for the Babun2 mockups folder.
# Mirrors serve.mjs: port 4321, serves ./mockups, "/" -> clients-current.html.
# Fallback runtime for machines without Node installed.
import os
from http.server import HTTPServer, SimpleHTTPRequestHandler

ROOT = os.path.dirname(os.path.abspath(__file__))
PORT = 4321


class Handler(SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    def translate_path(self, path):
        # Map "/" to the clients mockup, then defer to default handling.
        if path in ("/", ""):
            path = "/clients-current.html"
        return super().translate_path(path)

    def end_headers(self):
        self.send_header("Cache-Control", "no-cache")
        super().end_headers()

    def log_message(self, fmt, *args):
        # Keep stdout quiet but visible for debugging.
        print("[serve.py] " + (fmt % args))


if __name__ == "__main__":
    print(f"Babun2 mockups server running at http://localhost:{PORT}/")
    HTTPServer(("127.0.0.1", PORT), Handler).serve_forever()
